serde/src/de/content.rs

```diff
@@ -0,0 +1,732 @@
+use core::fmt;
+use core::marker::PhantomData;
+
+#[cfg(all(not(feature = "std"), feature = "collections"))]
+use collections::{String, Vec};
+
+#[cfg(all(feature = "alloc", not(feature = "std")))]
+use alloc::boxed::Box;
+
+use de::{
+    self,
+    Deserialize,
+    DeserializeSeed,
+    Deserializer,
+    Visitor,
+    SeqVisitor,
+    MapVisitor,
+    EnumVisitor,
+};
+
+/// Used from generated code to buffer the contents of the Deserializer when
+/// deserializing untagged enums and internally tagged enums.
+///
+/// Not public API. Use serde-value instead.
+#[allow(missing_docs)]
+#[derive(Debug)]
+pub enum Content<E> {
+    // Don't mind the PhantomData, just need to use E somewhere.
+    Bool(bool, PhantomData<E>),
+
+    U8(u8),
+    U16(u16),
+    U32(u32),
+    U64(u64),
+
+    I8(i8),
+    I16(i16),
+    I32(i32),
+    I64(i64),
+
+    F32(f32),
+    F64(f64),
+
+    Char(char),
+    String(String),
+    Bytes(Vec<u8>),
+
+    None,
+    Some(Box<Content<E>>),
+
+    Unit,
+    Newtype(Box<Content<E>>),
+    Seq(Vec<Content<E>>),
+    Map(Vec<(Content<E>, Content<E>)>),
+}
+
+impl<E> Deserialize for Content<E> {
+    fn deserialize<D: Deserializer>(deserializer: D) -> Result<Self, D::Error> {
+        // Untagged and internally tagged enums are only supported in
+        // self-describing formats.
```

Comment:
We could find some way to support other formats through some rewinding API. This would also have uses for streaming.